File: src/Control.cpp
#include "Control.h"
#include "ErrorExit.h"
#include "../3rdparty/tracy/Tracy.hpp"
#include "ServiceLocator.h"
#include "ResourcesD2D1.h"
#include <d2d1.h>

bool mj::Control::TranslateClientPoint(int16_t* pX, int16_t* pY)
{
  MJ_UNINITIALIZED POINT point;
  point.x = *pX;
  point.y = *pY;

  MJ_UNINITIALIZED RECT rect;
  rect.left   = this->xParent;
  rect.right  = this->xParent + this->width;
  rect.top    = this->yParent;
  rect.bottom = this->yParent + this->height;

  if (::PtInRect(&rect, point))
  {
    *pX -= this->xParent;
    *pY -= this->yParent;
    return true;
  }

  return false;
}

void mj::Control::OnPaint(ID2D1RenderTarget* pRenderTarget)
{
  ZoneScoped;

  MJ_UNINITIALIZED D2D1_MATRIX_3X2_F transform;
  pRenderTarget->GetTransform(&transform);
  pRenderTarget->SetTransform(D2D1::Matrix3x2F::Translation(D2D1::SizeF(this->xParent, this->yParent)) * transform);
  MJ_DEFER(pRenderTarget->SetTransform(&transform));

  D2D1_RECT_F rect = D2D1::RectF(0, 0, this->width, this->height);
  pRenderTarget->PushAxisAlignedClip(rect, D2D1_ANTIALIAS_MODE_ALIASED);
  MJ_DEFER(pRenderTarget->PopAxisAlignedClip());

  pRenderTarget->FillRectangle(rect, res::d2d1::Brush());

  D2D1_ANTIALIAS_MODE antialiasMode = pRenderTarget->GetAntialiasMode();
  pRenderTarget->SetAntialiasMode(D2D1_ANTIALIAS_MODE_ALIASED);
  MJ_DEFER(pRenderTarget->SetAntialiasMode(antialiasMode));

  this->Paint(pRenderTarget);
}

File: zig/src/control.zig
usingnamespace @import("win32.zig");
const d2d1 = @import("d2d1.zig");

pub fn Control() type {
    return struct {
        const Self = @This();

        /// <summary>
        /// X position relative to parent control.
        /// Note: Use zero if you want to specify the left edge inside the Paint() method.
        /// </summary>
        xParent: i16,

        /// <summary>
        /// Y position relative to parent control.
        /// Note: Use zero if you want to specify the top edge inside the Paint() method.
        /// </summary>
        yParent: i16,
        width: i16,
        height: i16,

        // Pure virtual
        initFn: fn (self: *Self, pAllocator: *AllocatorBase) void,
        resetFn: fn (self: *Self, pRenderTarget: *ID2D1RenderTarget) void,
        destroyFn: fn (self: *Self) void,

        // Virtual
        onMouseMoveFn: fn (self: *Self, x: i16, y: i16) void,
        onDoubleClickFn: fn (self: *Self, x: i16, y: i16, mkMask: u16) void,
        onMouseWheelFn: fn (self: *Self, x: i16, y: i16, mkMask: u16, zDelta: i16) void,
        onContextMenuFn: fn (self: *Self, clientX: i16, clientY: i16, screenX: i16, screenY: i16) void,
        onSizeFn: fn (self: *Self, clientX: i16, clientY: i16, screenX: i16, screenY: i16) void,
        onLeftButtonDownFn: fn (self: *Self, x: i16, y: i16) bool,
        onLeftButtonUpFn: fn (self: *Self, x: i16, y: i16) bool,

        pub fn OnPaint(pRenderTarget: *ID2D1RenderTarget) void {
            // ZoneScoped;
            var transform: D2D1_MATRIX_3X2_F = undefined;
            ID2D1RenderTarget_GetTransform(pRenderTarget, &transform);
            ID2D1RenderTarget_SetTransform(d2d1.mul(d2d1.translation(self.xParent, self.yParent), transform));
            defer ID2D1RenderTarget_SetTransform(pRenderTarget, &transform);

            ID2D1RenderTarget_PushAxisAlignedClip(pRenderTarget, D2D1_RECT_F{
                .left = 0.0,
                .top = 0.0,
                .right = self.width,
                .bottom = self.height,
            }, D2D1_ANTIALIAS_MODE_ALIASED);
            defer ID2D1RenderTarget_PopAxisAlignedClip();

            const antialiasMode = ID2D1RenderTarget_GetAntialiasMode();
            ID2D1RenderTarget_SetAntialiasMode(D2D1_ANTIALIAS_MODE_ALIASED);
            defer ID2D1RenderTarget_SetAntialiasMode(antialiasMode);

            self.Paint(pRenderTarget);
        }

        pub fn init(self: *Self, pAllocator: *AllocatorBase) void {
            return self.nextFn(self, pAllocator);
        }

        pub fn reset(self: *Self, pRenderTarget: *ID2D1RenderTarget) void {
            return self.nextFn(self, pRenderTarget);
        }

        pub fn destroy(self: *Self) void {
            return self.nextFn(self);
        }
    };
}
